src/mmu/timers.rs

```diff
@@ -0,0 +1,292 @@
+            tima_write_lock_counter: 0,
+            tima_falling_edge_detected: false,
+        }
+    }
+}
+
+impl Mmu {
+    // One tick is 1 t-cycle
+    pub fn tick_timers(&mut self) {
+        self.timers.system_clock_counter += 1;
+        self.timers.tima_write_lock_counter = self.timers.tima_write_lock_counter.saturating_sub(1);
+
+        // The system clock is a 16-bit number in t-cycles, but it is only incremented once every m-cycle.
+        if self.timers.system_clock_counter == T_CYCLES_PER_M_CYCLE {
+            self.timers.system_clock_counter = 0;
+            self.timers.system_clock = self.timers.system_clock.wrapping_add(T_CYCLES_PER_M_CYCLE);
+        }
+
+        // ----- Updating DIV
+        let div_value = (self.timers.system_clock >> 8) as u8;
+        self.write_byte_override(DIV_ADDR, div_value);
+
+        // ----- Updating TIMA
+        let tac_enable = self.get_tac_enable();
+        let tima_bit_is_active = self.get_tima_bit_state(false);
+        let tima_bit_was_active = self.get_tima_bit_state(true);
+
+        // TIMA's special overflow behavior occurs the cycle AFTER an overflow is detected.
+        if self.timers.tima_overflowed {
+            self.process_tima_overflow();
+        }
+
+        // TIMA increments while enabled, and on a falling clock edge. The second condition is
+        // there because memory writes can trigger a clock edge detection that occurs elsewhere.
+        if tac_enable && tima_bit_was_active && !tima_bit_is_active
+            || self.timers.tima_falling_edge_detected
+        {
+            self.increment_tima();
+            self.timers.tima_falling_edge_detected = false;
+        }
+
+        // It's important to update the previous clockstate here, instead of at the beginning of the loop.
+        // This is because the system clock can be reset if something writes to the div timer.
+        // Therefore, the system clock state might be different by the next time this function is called.
+        self.timers.system_clock_prev = self.timers.system_clock;
+    }
+
+    fn increment_tima(&mut self) {
+        let mut tima_byte = self.read_byte(TIMA_ADDR);
+        let overflowed;
+
+        (tima_byte, overflowed) = tima_byte.overflowing_add(1);
+
+        self.write_byte_override(TIMA_ADDR, tima_byte);
+
+        if overflowed {
+            self.timers.tima_overflowed = true;
+        }
+    }
+
+    /// Overflows of TIMA send a timer interrupt. When TIMA overflows, its value is set
+    /// to the value stored in TMA.
+    fn process_tima_overflow(&mut self) {
+        self.timers.tima_overflowed = false;
+        self.timers.tima_write_lock_counter = TIMA_WRITE_LOCK_T_CYCLES;
+
+        self.request_interrupt(TIMER_INTERRUPT_BIT);
+        let tma_value = self.read_byte(TMA_ADDR);
+        self.write_byte_override(TIMA_ADDR, tma_value);
+    }
+
+    fn get_tac_enable(&self) -> bool {
+        let byte = self.read_byte(TAC_ADDR);
+        get_bit(byte, TAC_ENABLE_BIT)
```

Comment:
this sequence gets repeated enough times in your code you might consider adding a utility function `self.get_bit(address, bit) -> bool`